File: src/main/java/me/sullivan/glasslang/parser/nodes/BinOpNode.java
package me.sullivan.glasslang.parser.nodes;

import me.sullivan.glasslang.interpreter.Interpreter;
import me.sullivan.glasslang.interpreter.primitives.types.Primitive;
import me.sullivan.glasslang.lexer.token.Token;

public class BinOpNode extends Node
{
    private Node left;
    private Node right;

    public BinOpNode(Node left, Token token, Node right)
    {
        super(token);
        this.left = left;
        this.right = right;
    }

    @Override
    public Token getValue()
    {
        return token;
    }

    @Override
    public String toString()
    {
        return "BinOp(" + left + ", " + getValue().getType() + ", " + right + ")";
    }

    public Node getLeft()
    {
        return left;
    }

    public Node getRight()
    {
        return right;
    }

    @Override
    public Primitive<?> visitor(Interpreter interpreter)
    {
        return interpreter.visitBinOp(this);
    }
}

File: src/me/sullivan/glasslang/parser/nodes/BinOpNode.kt
package me.sullivan.glasslang.parser.nodes

import me.sullivan.glasslang.interpreter.Interpreter
import me.sullivan.glasslang.interpreter.primitives.Primitive
import me.sullivan.glasslang.lexer.token.Token

class BinOpNode(val left: Node, token: Token, val right: Node) : Node(token) {

    override val value: Token
        get() = token

    override fun toString(): String {
        return "BinOp($left, ${value.type}, $right)"
    }

    override fun visitor(interpreter: Interpreter): Primitive<*> {
        return interpreter.visitBinOp(this)
    }
}
